### PR TITLE
Reorder arguments for calloc()-like functions, part #2

### PR DESCRIPTION
**Pull request by mrc0mmand**
To appease gcc-14's -Wcalloc-transposed-args check.

Follow-up for https://github.com/systemd/systemd/commit/2a9ab0974bb290bc66dc84d909c33d23361b0752.

---

A couple of hours ago gcc-14 was [pushed](https://bodhi.fedoraproject.org/updates/FEDORA-2024-286f139579) to Rawhide and some Packit jobs already picked it up and started failing. This PR should take the edge off, but there's still bunch of warnings when building with -O2 which will need further inspection.

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -341,6 +341,7 @@ endif
 basic_disabled_warnings = [
         '-Wno-missing-field-initializers',
         '-Wno-unused-parameter',
+        '-Wno-nonnull-compare',
```
Comment (poettering) [inline meson.build:344]: hmm, at which of our calls does that actually trigger?

**Comment by poettering** [comment]
lgtm

**Comment by poettering** [comment]
i marked it green, but you marked it draft. just means: go ahead and merge it if you think this is ready

File: meson.build (hunk 1 above)
Comment (mrc0mmand) [reply meson.build:344]: This check complains everywhere where we use `int x[static X]` in function declaration and then call `assert()` on it:

```
../src/basic/fd-util.h: In function ‘format_proc_fd_path’:
../src/fundamental/macro-fundamental.h:71:41: warning: ‘nonnull’ argument ‘buf’ compared to NULL [-Wnonnull-compare]
   71 | #define _unlikely_(x) (__builtin_expect(!!(x), 0))
      |                                         ^~~~~
../src/basic/macro.h:148:21: note: in expansion of macro ‘_unlikely_’
  148 |                 if (_unlikely_(!(expr)))                                \
      |                     ^~~~~~~~~~
../src/basic/macro.h:166:22: note: in expansion of macro ‘assert_message_se’
  166 | #define assert(expr) assert_message_se(expr, #expr)
      |                      ^~~~~~~~~~~~~~~~~
../src/basic/fd-util.h:129:9: note: in expansion of macro ‘assert’
  129 |         assert(buf);
      |         ^~~~~~
```

https://github.com/systemd/systemd/blob/02b346d64f3667ebb857f064a3e8ad84682d0e5f/src/basic/fd-util.h#L128-L133

It's actually quite a lot of warnings, so I'm not completely sure if we want drop the asserts or disable the warning just for the assert stuff via using the _Pragma() diagnostic push stuff.

**Comment by mrc0mmand** [comment]
Ugh, there's one warning which I can't really find a way around:

```
../src/basic/macro.h:395:90: warning: array subscript 1 is outside array bounds of ‘uint8_t[1]’ {aka ‘unsigned char[1]’} [-Warray-bounds=]
  395 |              ((long)(_current_ - _entries_) < (long)ELEMENTSOF(_entries_)) && ({ entry = *_current_; true; }); \
../src/basic/macro.h:392:9: note: in expansion of macro ‘_VA_ARGS_FOREACH’
  392 |         _VA_ARGS_FOREACH(entry, UNIQ_T(_entries_, UNIQ), UNIQ_T(_current_, UNIQ), ##__VA_ARGS__)
      |         ^~~~~~~~~~~~~~~~
../src/test/test-macro.c:328:9: note: in expansion of macro ‘VA_ARGS_FOREACH’
  328 |         VA_ARGS_FOREACH(u8, 0xff) {
      |         ^~~~~~~~~~~~~~~
../src/fundamental/macro-fundamental.h:163:37: note: at offset 1 into object ‘__unique_prefix__entries_183’ of size 1
  163 | #define UNIQ_T(x, uniq) CONCATENATE(__unique_prefix_, CONCATENATE(x, uniq))
      |                                     ^~~~~~~~~~~~~~~~
../src/basic/macro.h:394:28: note: in definition of macro ‘_VA_ARGS_FOREACH’
  394 |         for (typeof(entry) _entries_[] = { __VA_ARGS__ }, *_current_ = _entries_; \
      |                            ^~~~~~~~~
../src/fundamental/macro-fundamental.h:109:27: note: in expansion of macro ‘XCONCATENATE’
  109 | #define CONCATENATE(x, y) XCONCATENATE(x, y)
      |                           ^~~~~~~~~~~~
../src/fundamental/macro-fundamental.h:163:25: note: in expansion of macro ‘CONCATENATE’
  163 | #define UNIQ_T(x, uniq) CONCATENATE(__unique_prefix_, CONCATENATE(x, uniq))
      |                         ^~~~~~~~~~~
../src/basic/macro.h:392:33: note: in expansion of macro ‘UNIQ_T’
  392 |         _VA_ARGS_FOREACH(entry, UNIQ_T(_entries_, UNIQ), UNIQ_T(_current_, UNIQ), ##__VA_ARGS__)
      |                                 ^~~~~~
../src/test/test-macro.c:328:9: note: in expansion of macro ‘VA_ARGS_FOREACH’
  328 |         VA_ARGS_FOREACH(u8, 0xff) {
      |         ^~~~~~~~~~~~~~~
[2414/2414] Generating export-dbus-interfaces with a custom command
```

In:
https://github.com/systemd/systemd/blob/1cdd8b1b66ec478ea2ed81ffa371e93f4b27c61e/src/basic/macro.h#L385-L391

I tried a couple of solutions, like using the first value as a sentinel, but that feels icky (and the macro can be used even without arguments). @poettering @yuwata any idea here?

**Comment by mrc0mmand** [comment]
I pushed my latest attempt on making gcc-14 happy with the `_VA_ARGS_FOREACH()` macro - it's not particularly pretty, but it makes gcc-14 shut up and the tests still pass.

**Review by bluca — APPROVED** [review]

**Comment by fbuihuu** [comment]
> May I ask why it has been backported to the v254 stable tree ?
> 
> This bump the requirement on the version of gcc from 4.7 to 8.x which is not ok for a stable release.
 
https://github.com/systemd/systemd-stable/issues/368

**Pull request merged**